metadata/src/main/java/org/apache/kafka/image/ClusterDelta.java

```diff
@@ -102,10 +104,34 @@ public void replay(UnfenceBrokerRecord record) {
     public void replay(BrokerRegistrationChangeRecord record) {
         BrokerRegistration broker =
             getBrokerOrThrow(record.brokerId(), record.brokerEpoch(), "change");
-        if (record.fenced() < 0) {
-            changedBrokers.put(record.brokerId(), Optional.of(broker.cloneWithFencing(false)));
-        } else if (record.fenced() > 0) {
-            changedBrokers.put(record.brokerId(), Optional.of(broker.cloneWithFencing(true)));
+        Optional<BrokerRegistrationFencingChange> fencingChange =
+            BrokerRegistrationFencingChange.fromValue(record.fenced());
+        if (!fencingChange.isPresent()) {
+            throw new IllegalStateException(String.format("Unable to replay %s: unknown " +
+                "value for fenced field: %d", record.toString(), record.fenced()));
+        }
+        Optional<BrokerRegistrationInControlledShutdownChange> inControlledShutdownChange =
+            BrokerRegistrationInControlledShutdownChange.fromValue(record.inControlledShutdown());
+        if (!inControlledShutdownChange.isPresent()) {
+            throw new IllegalStateException(String.format("Unable to replay %s: unknown " +
+                "value for inControlledShutdown field: %d", record.toString(), record.inControlledShutdown()));
+        }
+
+        replayRegistration(record.brokerId(), broker, fencingChange.get(), inControlledShutdownChange.get());
+    }
+
+    private void replayRegistration(
+        int brokerId,
+        BrokerRegistration broker,
+        BrokerRegistrationFencingChange fencingChange,
+        BrokerRegistrationInControlledShutdownChange inControlledShutdownChange
+    ) {
+        if (fencingChange != BrokerRegistrationFencingChange.NONE
+            || inControlledShutdownChange != BrokerRegistrationInControlledShutdownChange.NONE) {
```

Comment:
If you follow my other comment regarding `cloneWith`, you may be able to replace this check with a check against the `BrokerRegistration` changing.

As we had more changeable registration field developers need to proliferate this kind of changes.